### PR TITLE
Fix compilation errors resulting from disabling impdep1

### PR DESCRIPTION
**Pull request by EricYangIBM**
Unused variable J9JavaVM *vm in swalk.c when building with OJDK MH
caused compilation error.

Fixes: #12705
Signed-off-by: Eric Yang <eric.yang@ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
@EricYangIBM Did you test these changes locally or through personal builds? Do they work?

**Comment by babsingh** [comment]
Also, please squash the two commits into a single commit once you have confirmed that your changes work.

**Comment by babsingh** [comment]
OJDK MH build cmds:
```
// Compile + test builds
jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

// Only compile builds
jenkins compile sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29
```

**Comment by gacholio** [comment]
This doesn't yet seem to exclude the MHInterprerer compiles.

**Comment by babsingh** [comment]
I think just excluding `include "MHInterpreter.hpp"` in `BytecodeInterpreter.hpp` should prevent its compilation. @EricYangIBM can you confirm this point?

**Comment by babsingh** [comment]
++ @gacholio senior reviewer.

**Comment by EricYangIBM** [comment]
> I think just excluding `include "MHInterpreter.hpp"` in `BytecodeInterpreter.hpp` should prevent its compilation. @EricYangIBM can you confirm this point?

@babsingh I think it is still compiling. I get an error in MHInterpreter.hpp about impdep1PC even with removing the #include

**Comment by gacholio** [comment]
The cpp file is failing, not the hpp. There are multiple copies being compiled.

**Comment by babsingh** [comment]
We can try surrounding `MHInterpreter` source code with `#if defined(J9VM_OPT_METHOD_HANDLE)`. The other approach will be to update the make files to stop building `MHInterpreter`.

**Comment by gacholio** [comment]
Let's go with the easy solution for now (ifdeffing the file). There was a problem on Mac regarding empty files, but I think that may have been resolved.

**Comment by keithc-ca** [comment]
Changes for DDR will also be required:
```
Missing required field: J9JavaVM.impdep1PC
```
See https://github.com/eclipse-openj9/openj9/blob/master/debugtools/DDR_VM/src/com/ibm/j9ddr/AuxFieldInfo29.dat#L247.
That line needs to change to
```
J9JavaVM.impdep1PC = U_8*
```
Also, the references in [`StackWalker.java`](https://github.com/eclipse-openj9/openj9/blob/master/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/stackwalker/StackWalker.java#L965) will need to handle `NoSuchFieldException`.

**Comment by gacholio** [comment]
Stackwalker code:
```
	if ((walkState.pc != vm.impdep1PC()) && (walkState.pc != (vm.impdep1PC().addOffset(3)))) {
		walkState.bytecodePCOffset = walkState.pc.sub(walkState.method.bytecodes().getAddress());	
	} else {
		walkState.bytecodePCOffset = U8Pointer.cast(new UDATA(0));
	}
```
This will need to be changed so that the true path is taken if the field is not found:
```
boolean isImpdep1 = false;
try {
   isImpdep1 = (walkState.pc == vm.impdep1PC()) || (walkState.pc == (vm.impdep1PC().addOffset(3)));
} catch (NoSuchFieldException e) {}
if (!isImpdep1) {
...
```
@keithc-ca Is this what you had in mind?

File: runtime/vm/MHInterpreter.inc [hunk 1]
```diff
@@ -19,6 +19,7 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
+#if defined(J9VM_OPT_METHOD_HANDLE)
```
Comment (gacholio) [inline runtime/vm/MHInterpreter.inc:22]: To be strictly correct, the ifdef should appear after including j9cfg.h.

Comment (keithc-ca) [reply runtime/vm/MHInterpreter.inc:22]: More than that, it's just wrong before that include (the symbol will always be undefined).

**Comment by keithc-ca** [comment]
Something like that, but the catch block deserves a comment. The behavior should match `swalk.c`.

**Comment by keithc-ca** [comment]
I also note that `U8Pointer.cast(new UDATA(0))` is a long-winded, and more expensive, way to get the same thing as `U8Pointer.NULL`.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/stackwalker/StackWalker.java [hunk 2]
```diff
@@ -962,10 +962,16 @@ private void walkBytecodeFrame(WalkState walkState)
 				J9JavaVMPointer vm = walkState.walkThread.javaVM();
 				
 				walkState.constantPool = ConstantPoolHelpers.J9_CP_FROM_METHOD(walkState.method);
-				if ((walkState.pc != vm.impdep1PC()) && (walkState.pc != (vm.impdep1PC().addOffset(3)))) {
+				boolean isImpdep1 = false;
+				try {
+					isImpdep1 = (walkState.pc == vm.impdep1PC()) || (walkState.pc == (vm.impdep1PC().addOffset(3)));
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/stackwalker/StackWalker.java:967]: Redundant parentheses in `(vm.impdep1PC().addOffset(3))`.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 3]
```diff
@@ -161,6 +164,7 @@ class INTERPRETER_CLASS
 	 * @param methodHandle[in] The MethodHandle to run
 	 * @return the next action to take.
 	 */
+#if defined(J9VM_OPT_METHOD_HANDLE)
```
Comment (keithc-ca) [inline runtime/vm/BytecodeInterpreter.hpp:167]: This should precede the comment for the function.

**Comment by EricYangIBM** [comment]
Building on my local machine was successful.

File: runtime/vm/MHInterpreter.inc (hunk 1 above)
Comment (gacholio) [reply runtime/vm/MHInterpreter.inc]: This isn't a real file - it's included from a .cpp that almost certainly already has included j9cfg.h.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 4]
```diff
@@ -41,11 +41,14 @@
 #include "jitregmap.h"
 #include "j2sever.h"
 #include "vmaccess.h"
+#include "objhelp.h"
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp:44]: Why has this been added?

Comment (keithc-ca) [reply runtime/vm/BytecodeInterpreter.hpp:44]: Some declarations came via `MHInterpreter.hpp` which may not be included with this change.

**Comment by keithc-ca** [comment]
I think we're ready for this to be squashed.

File: runtime/vm/MHInterpreter.inc (hunk 1 above)
Comment (keithc-ca) [reply runtime/vm/MHInterpreter.inc]: Perhaps, but better safe than sorry.

File: runtime/vm/MHInterpreter.inc [hunk 5]
```diff
@@ -2558,3 +2559,4 @@ ffi_exit:
 	return ret;
 }
 #endif /* J9VM_OPT_PANAMA */
+#endif /* defined(J9VM_OPT_METHOD_HANDLE) */
```
Comment (gacholio) [inline runtime/vm/MHInterpreter.inc:2562]: Does this file end with a newline?

Comment (EricYangIBM) [reply runtime/vm/MHInterpreter.inc:2562]: Yes.
image

**Comment by EricYangIBM** [comment]
Commits are squashed

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk8

**Comment by gacholio** [comment]
jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by EricYangIBM** [comment]
Builds failed:
```
[2021-05-18T19:41:15.637Z] /home/jenkins/workspace/Build_JDK16_x86-64_linux_ojdk292_Personal/build/linux-x86_64-server-release/support/gensrc/openj9.dtfj/com/ibm/j9ddr/vm29/j9/gc/GCObjectHeapIteratorAddressOrderedList_V2.java:42: error: exception NoSuchFieldException is never thrown in body of corresponding try statement
[2021-05-18T19:41:15.637Z] 		} catch (NoSuchFieldException e) {
[2021-05-18T19:41:15.637Z] 		  ^
```
```
[2021-05-18T19:41:15.637Z] /home/jenkins/workspace/Build_JDK16_x86-64_linux_ojdk292_Personal/build/linux-x86_64-server-release/support/gensrc/openj9.dtfj/com/ibm/j9ddr/vm29/j9/stackwalker/StackWalker.java:967: error: cannot find symbol
[2021-05-18T19:41:15.637Z] 					isImpdep1 = (walkState.pc == vm.impdep1PC()) || (walkState.pc == vm.impdep1PC().addOffset(3));
[2021-05-18T19:41:15.637Z] 					                               ^
[2021-05-18T19:41:15.637Z]   symbol:   method impdep1PC()
[2021-05-18T19:41:15.637Z]   location: variable vm of type J9JavaVMPointer
```
I am confused as how the fields are generated.

**Comment by gacholio** [comment]
Hopefully @keithc-ca knows what to do. Is the catch issue that this should be an Error and not an Exception?

**Comment by keithc-ca** [comment]
The failure is because ibmruntimes/openj9-openjdk-jdk16#29 does not include ibmruntimes/openj9-openjdk-jdk16#50. I pushed a merge commit to fix that.

**Comment by keithc-ca** [comment]
Jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by keithc-ca** [comment]
This failure seems relevant:
```
[2021-05-19T15:45:25.671Z] FAILED: test_RevealDirect_TargetVarargsAdaptor
[2021-05-19T15:45:25.671Z] java.lang.AssertionError: 
[2021-05-19T15:45:25.671Z] 	at org.testng.AssertJUnit.fail(AssertJUnit.java:59) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:33) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.openj9.test.java.lang.invoke.Test_MethodHandleInfo.test_RevealDirect_TargetVarargsAdaptor(Test_MethodHandleInfo.java:667) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/jvmtest/functional/Java8andUp/GeneralTest.jar)
[2021-05-19T15:45:25.671Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) from jrt:/java.base
[2021-05-19T15:45:25.671Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78) from jrt:/java.base
[2021-05-19T15:45:25.671Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) from jrt:/java.base
[2021-05-19T15:45:25.671Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:567) from jrt:/java.base
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:580) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.TestRunner.privateRun(TestRunner.java:648) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.TestRunner.run(TestRunner.java:505) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.SuiteRunner.run(SuiteRunner.java:364) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.671Z] 	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.runSuites(TestNG.java:1049) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.run(TestNG.java:1017) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.privateMain(TestNG.java:1354) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 	at org.testng.TestNG.main(TestNG.java:1323) from jdk.internal.loader.ClassLoaders$AppClassLoader@4e8e2499(file:/F:/Users/jenkins/workspace/Test_openjdk16_j9_sanity.functional_x86-64_windows_ojdk292_Personal_testList_0/openjdk-tests/TKG/lib/testng.jar)
[2021-05-19T15:45:25.672Z] 
[2021-05-19T15:45:25.672Z] 
[2021-05-19T15:45:25.672Z] ===============================================
[2021-05-19T15:45:25.672Z]     JCL_TEST_Java-Lang-Invoke
[2021-05-19T15:45:25.672Z]     Tests run: 31, Failures: 1, Skips: 0
[2021-05-19T15:45:25.672Z] ===============================================
```

**Comment by gacholio** [comment]
I don't think this PR could cause any real test failures in the new MH builds, but @babsingh can confirm if all the failures are known (why haven't these been excluded?).

**Comment by babsingh** [comment]
@gacholio Yes, all the failures are known. PRs (https://github.com/eclipse-openj9/openj9/pull/12716 & https://github.com/eclipse-openj9/openj9/pull/12668) have been created to resolve some of the failures. We plan to disable tests where the test behaviour doesn't check for the correct RI behaviour. Recently, I had a discussion with the test team on disabling tests, https://github.com/eclipse-openj9/openj9/issues/11935#issuecomment-839873698. I will create a PR to disable tests soon.

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**